### DISCUSSION TITLE
Deta not detecting module in `requirements.txt`

### DISCUSSION THREAD

**Discussion by eRuaro** [deta/cloud-docs#327] [Help]
I'm currently trying to deploy an API which uses the Twilio SDK with the following details.
```
{
        "name": "project-name",
        "id": "project-id",
        "project": "default",
        "runtime": "python3.9",
        "endpoint": "project-endpoint",
        "region": "ap-southeast-1",
        "dependencies": [
                "fastapi==0.78.0",
                "uvicorn==0.18.2",
                "deta==1.1.0",
                "twilio==7.11.0"
        ],
        "environment_variables": [
                "TWILIO_ACCOUNT_SID",
                "TWILIO_AUTH_TOKEN"
        ],
        "visor": "disabled",
        "http_auth": "disabled"
}
```

As you can see, Twilio is part of the dependencies, and the environment variables as well have Twilio authentication keys in them, however I'm getting this error in my logs.
```
[ERROR] ModuleNotFoundError: No module named 'twilio'
***
***
***
twilio.base.exceptions.TwilioRestException: HTTP 401 error: Unable to create record: Authentication Error - invalid username
```

How do I fix this? Shouldn't twilio be found since it's part of the dependencies? For the authentication error, I believe it's just a matter of not having the twilio module in the deployed micro as the docs said that I should just include the environment variables.

**Comment by jnsougata** [comment]
1. you won't be needing `uvicorn` for running the app in micro
2. as program raised `twilio.base.exceptions.TwilioRestException` this I believe there's something wrong with your code!
looking forward to know more about your issue to help you.

**Reply by eRuaro** [reply]
Here's the snippet of code that uses the twilio SDK. 
```
from twilio.rest import Client
from fastapi import FastAPI, status, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from uvicorn import run

app = FastAPI()

origins = ["*"]
methods = ["*"]
headers = ["*"]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=methods,
    allow_headers=headers
)

@app.post("/message/send", status_code = status.HTTP_201_CREATED)
async def post_message(toNumber: str, fromNumber: str, message: str):
    # error handling stuffs below, won't be including it here

    # start of code snippet using twilio sdk
    account_sid = os.environ["TWILIO_ACCOUNT_SID"]
    auth_token = os.environ["TWILIO_AUTH_TOKEN"]
    client = Client(account_sid, auth_token) # Client comes from twilio.rest

    clientMessage = client.messages.create(
        body=message,
        to=toNumber,
        from_=fromNumber,
    )

    return {
        "toNumber": toNumber,
        "fromNumber": fromNumber,
        "message": message,
        "messageBody": clientMessage.body
    }

if __name__ == "__main__":
    port = int(os.environ.get('PORT', 5000))
    run(app, host="0.0.0.0", port=port)
```

In my localhost, it works correctly. I think the issue at hand is that Deta fails to collect the values for the `account_sid` and `auth_token` from the `.env` file. I ran `deta update -e .env` to get it from the `.env` file which looks like this.
```
TWILIO_ACCOUNT_SID='my-sid'
TWILIO_AUTH_TOKEN='my-token'
```

**Reply by eRuaro** [reply]
I edited the above code to have the `python-dotenv` library to access the `.env` file but I'm still running into the twilio authentication error.

Here's the sequence I did. I first added `python-dotenv` to my `requirements.txt` file.
```
// in requirements.txt
fastapi==0.78.0
uvicorn==0.18.2
deta==1.1.0
twilio==7.11.0
python-dotenv==0.20.0
```

And ran with `pip3 install -r requirements.txt`. After this, I edited the source code file to use the `python-dotenv` module.
```
from twilio.rest import Client
from fastapi import FastAPI, status, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from uvicorn import run
from dotenv import load_dotenv

load_dotenv()

app = FastAPI()

origins = ["*"]
methods = ["*"]
headers = ["*"]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=methods,
    allow_headers=headers
)

@app.post("/message/send", status_code = status.HTTP_201_CREATED)
async def post_message(toNumber: str, fromNumber: str, message: str):
    # error handling stuffs below, won't be including it here

    # start of code snippet using twilio sdk
    account_sid = os.getenv("TWILIO_ACCOUNT_SID")
    auth_token = os.getenv("TWILIO_AUTH_TOKEN")
    client = Client(account_sid, auth_token) # Client comes from twilio.rest

    clientMessage = client.messages.create(
        body=message,
        to=toNumber,
        from_=fromNumber,
    )

    return {
        "toNumber": toNumber,
        "fromNumber": fromNumber,
        "message": message,
        "messageBody": clientMessage.body
    }

if __name__ == "__main__":
    port = int(os.environ.get('PORT', 5000))
    run(app, host="0.0.0.0", port=port)
```
I then updated my micro by running `deta deploy` and updated the `.env` file in the micro, even though I didn't edit the `.env` file with `deta update -e .env`.

Running tests, and the app locally, I find that everything is working fine. On Deta however, I'm still getting this error.
```
twilio.base.exceptions.TwilioRestException: HTTP 401 error: Unable to create record: Authentication Error - invalid username
```

Is there a way I can debug the environment variables in the micro?

**Reply by abdelhai** [reply]
just to confirm, you did run `deta update -e .env`? if no, then please do.
i can see that we have the env vars set in your micro. to make sure the values are correct, you could print them out with pythons print function.

**Reply by eRuaro** [reply]
Hi, yup! I just ran, and it is still having the error. On localhost, it can read the environment variables. On Deta however, I think it's not being read, or the returned value is `None`.
